# Выбор локации для скважины

# Well Location Selection


Допустим, вы работаете в добывающей компании «ГлавРосГосНефть». Нужно решить, где бурить новую скважину.

Вам предоставлены пробы нефти в трёх регионах: в каждом 10 000 месторождений, где измерили качество нефти и объём её запасов. Постройте модель машинного обучения, которая поможет определить регион, где добыча принесёт наибольшую прибыль. Проанализируйте возможную прибыль и риски техникой *Bootstrap.*

Шаги для выбора локации:

- В избранном регионе ищут месторождения, для каждого определяют значения признаков;
- Строят модель и оценивают объём запасов;
- Выбирают месторождения с самыми высокими оценками значений. Количество месторождений зависит от бюджета компании и стоимости разработки одной скважины;
- Прибыль равна суммарной прибыли отобранных месторождений.

---


Suppose you work for the oil extraction company **GlavRosGosNeft**. The task is to determine where to drill a new well.

You are given oil samples from **three regions**. Each region contains **10,000 deposits**, where oil quality and reserve volume have been measured. Build a machine learning model to help identify the region where oil production will generate the highest profit. Analyze the potential profit and risks using the *Bootstrap* technique.

Steps for selecting a location:

- In the selected region, deposits are explored and feature values are determined for each of them;
- A model is built to estimate the reserve volume;
- Deposits with the highest predicted values are selected. The number of deposits depends on the company’s budget and the cost of developing one well;
- Profit is calculated as the total profit from the selected deposits.

#### Цель проекта:
- Построить модель для определения региона, где добыча принесёт наибольшую прибыль

#### Данные:
- В качестве данных для исследования даны файлы геологоразведки трёх регионов.

#### Ход исследования:
- Загрузка и подготовка данных:
  - Импортируем библиотеки и модули, необходимые для исследования.
  - Читаем файлы и сохраняем их в переменные.
  - Оцениваем содержимое датасетов и описываем его.
  - Обрабатываем данные.
- Обучение и проверка модели:
  - Напишем функцию для построения модели на данных, включающую все необходимые этапы обучения.
  - Применим эту функцию ко всем датасетам и сохраним результаты в переменные.
  - Из результатов по регионам извлечём нужные значения и сохраним их в отдельные переменные.
  - Выведем на экран средний запас предсказанного сырья и RMSE модели по регионам. Проанализируем результаты.
- Подготовка к расчёту прибыли:
  - Сохраним все ключевые значения для расчётов в отдельные переменные.
  - Рассчитаем достаточный объём сырья для безубыточной разработки.
  - Сравним объём сырья для безубыточной работы со средними запасами по регионам.
- Расчёт прибыли и рисков:
  - Напишем функцию для подсчёта прибыли и рисков.
  - Применим эту функцию поочерёдно к регионам и сохраним возвращённые значения в соответствующие переменные.
  - Из результатов по регионам извлечём нужные значения и сохраним их отдельно.
  - Выведем на экран для каждого региона:
    - среднюю прибыль
    - 95%-й доверительный интервал
    - риск убытков
  - Проанализируем результаты.
  - Выберем регион с наибольшей средней прибылью среди регионов с вероятностью убытков меньше 2.5%.
- Общий вывод.

---

#### Project Objective:
- To build a model for identifying the region where oil extraction will generate the highest profit.

#### Data:
- The study uses geological exploration datasets from three different regions.

#### Research Plan:
- Data loading and preparation:
  - Import the libraries and modules required for the analysis.
  - Read the files and store them in variables.
  - Examine and describe the contents of the datasets.
  - Preprocess the data.
- Model training and evaluation:
  - Create a function for building a model on the data, including all necessary training steps.
  - Apply this function to all datasets and store the results in variables.
  - Extract the relevant values from the regional results and store them in separate variables.
  - Print the average predicted reserve volume and the model’s RMSE for each region. Analyze the results.
- Preparation for profit calculation:
  - Store all key parameters for the calculations in separate variables.
  - Calculate the minimum reserve volume required for break-even development.
  - Compare the break-even reserve volume with the average reserves in each region.
- Profit and risk calculation:
  - Create a function to calculate profit and risks.
  - Apply this function to each region and store the returned values in separate variables.
  - Extract the necessary values from the regional results and save them separately.
  - Print for each region:
    - average profit
    - 95% confidence interval
    - risk of losses
  - Analyze the results.
  - Select the region with the highest average profit among those with a loss probability below 2.5%.
- Overall conclusion.

## Загрузка и подготовка данных

#### Импортируем библиотеки и модули:

In [2]:
#Собираем все импорты в одном месте
import pandas as pd
import numpy as np
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error
from sklearn.preprocessing import StandardScaler

#### Прочитаем файлы и сохраним их в переменные:

In [3]:
#локально
try:
    df_geo_1 = pd.read_csv('/Users/danielschollenberg/Desktop/geo_data_0.csv')
    df_geo_2 = pd.read_csv('/Users/danielschollenberg/Desktop/geo_data_1.csv')
    df_geo_3 = pd.read_csv('/Users/danielschollenberg/Desktop/geo_data_2.csv')
#с сервера
except:
    df_geo_1 = pd.read_csv('https://code.s3.yandex.net/datasets/geo_data_0.csv')
    df_geo_2 = pd.read_csv('https://code.s3.yandex.net/datasets/geo_data_1.csv')
    df_geo_3 = pd.read_csv('https://code.s3.yandex.net/datasets/geo_data_2.csv')

#### Оценим содержимое датасетов и опишем его:

In [4]:
print(df_geo_1.info())
df_geo_1.head(2)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB
None


,id,f0,f1,f2,product
0,txEyH,0.705745,-0.497823,1.22117,105.280062
1,2acmU,1.334711,-0.340164,4.36508,73.037750


In [5]:
print(df_geo_2.info())
df_geo_2.head(2)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB
None


,id,f0,f1,f2,product
0,kBEdx,-15.001348,-8.276000,-0.005876,3.179103
1,62mP7,14.272088,-3.475083,0.999183,26.953261


In [6]:
print(df_geo_3.info())
df_geo_3.head(2)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB
None


,id,f0,f1,f2,product
0,fwXo0,-1.146987,0.963328,-0.828965,27.758673
1,WJtFt,0.262778,0.269839,-2.530187,56.069697


##### Вывод:
- Содержимое датасетов идентично, это:
   - id — уникальный идентификатор скважины;
   - f0, f1, f2 — три признака точек;
   - product — объём запасов в скважине (тыс. баррелей).
- Столбец product является целевый признаком - target.
- Уникальный идентификатор скважины - столбец id, - не несет информативности для модели - его надо будет удалить.

##### Обработаем данные:

In [7]:
#Удалим столбец id во всех датасетах
df_geo_1 = df_geo_1.drop(['id'], axis=1)
df_geo_2 = df_geo_2.drop(['id'], axis=1)
df_geo_3 = df_geo_3.drop(['id'], axis=1)

In [8]:
#Оценим что вышло
display(df_geo_1.head(2))
display(df_geo_2.head(2))
df_geo_3.head(2)

,f0,f1,f2,product
0,0.705745,-0.497823,1.22117,105.280062
1,1.334711,-0.340164,4.36508,73.037750


,f0,f1,f2,product
0,-15.001348,-8.276000,-0.005876,3.179103
1,14.272088,-3.475083,0.999183,26.953261


,f0,f1,f2,product
0,-1.146987,0.963328,-0.828965,27.758673
1,0.262778,0.269839,-2.530187,56.069697


## Обучение и проверка модели

#### Объединять ли датасеты регионов месторождений в один для обучения модели или обучать модель на каждом регионе индивидуально?

- В нашем случае, предполагаю, что каждый регион представляет собой разные геологические особенности, а также разные характеристики месторождений, и считаю, что наилучшим подходом будет обучение модели на каждом регионе индивидуально.

- Обучение модели на каждом регионе отдельно позволит модели лучше улавливать уникальные особенности каждого региона и адаптироваться к ним. Это может привести к более точным прогнозам для каждого региона.

#### Напишем функцию model_training_and_validation:
- Которая будет иметь на вход аргумент data - с данными.
- Разбивать данные по признакам.
- Разбивать данные признаков на обучающую и валидационную выборки.
- Проводить масштабирование признаков. 
   - Хоть в задание ничего про это не сказано, и мы не знаем какие конкретно признаки имеют месторождения, однако их масштаб может отличаться. Например это могут быть признаки глубины месторождения и возраст его возникновения, маловероятно, что признаки определяющие запасы месторождений имеют одинаковый диапазон масштабов, ведь, вероятнее всего, обозначают соверешенно разные вещи. 
   - P.S: я увидел, при дальнейшим написание проекта, что это особо не влияет, но решил оставить.
- Создавать экземпляр модели, обучать его на обучающий выборке.
- Получать предсказания на на валидационной выборке.
- Вычислять среднее значение предсказанных запасов.
- Вычислять RMSE модели.
- Возвращать обученную на тренировочной выборке модель, значения предсказаний модели и правильных ответов на валидационной выборке, средний запас предсказанного сырья и RMSE модели.

In [9]:
def model_training_and_validation(data):
    #Разобьем данные по признакам
    features = data.drop(['product'], axis=1)
    target = data['product']
    #Разобъем данные признаков на обучающую и валидационную выборки
    features_train, features_valid, target_train, target_valid = train_test_split(
                                                                            features,
                                                                            target,
                                                                            test_size=0.75,
                                                                            random_state=12345
                                                                            )
    
    #Проведем масштабирование признаков с помощью StandardScaler
    #cоздадим экземпляр scaler
    scaler = StandardScaler()
    #обучим на тренировочной выборке
    scaler.fit(features_train)
    #масштабируем тренировочную выборку
    features_train = scaler.transform(features_train)
    #масштабируем тестовую выборку
    features_valid = scaler.transform(features_valid)
    
    #Создадим экземпляр модели
    model = LinearRegression()
    #Обучим модель на обучающей выборке
    model.fit(features_train, target_train)
    #Получим и сохраним в переменную предсказания модели на валидационной выборке
    predictions_valid = model.predict(features_valid)
    
    #Вычислим среднее значение предсказанных запасов
    mean_predictions_valid_volume = np.mean(predictions_valid)
    
    #Получим среднеквадратическую ошибку (MSE)
    mse = mean_squared_error(target_valid, predictions_valid)
    #Вычислить RMSE, вычислив квадратный корень из MSE
    rmse = np.sqrt(mse)
    
    #Вернем обученную на тренировочной выборке модель
    #Вернем значения предсказаний модели и правильных ответов на валидационной выборке 
    #Вернем средний запас предсказанного сырья и RMSE
    return model, predictions_valid, target_valid, mean_predictions_valid_volume, rmse

#### Применим функцию поочередно к датасетам регионов и сохраним возвращенные ею значения в соответствующии переменные.

In [10]:
#Применим функцию к каждому датасету и запишем выводимые ее значения в переменную для каждого датасета
result_model_geo_1 = model_training_and_validation(df_geo_1)
result_model_geo_2 = model_training_and_validation(df_geo_2)
result_model_geo_3 = model_training_and_validation(df_geo_3)

#### Из переменных с результатами работы функции по регионам вытащим каждое значения и запишим в соответствующую переменную

In [11]:
#Вытащим и запишим в переменные выводимые функцией значения для региона №1
model_geo_1 = result_model_geo_1[0]                          #Обученная модель для региона 
predictions_valid_geo_1 = result_model_geo_1[1]              #Предсказания модели на валидационной выборке для региона 
target_valid_geo_1 = result_model_geo_1[2]                   #Правильные ответы валидационной выборке для региона 
mean_predictions_valid_volume_geo_1 = result_model_geo_1[3]  #Средний запас предсказанного сырья для региона
rmse_geo_1 = result_model_geo_1[4]                           #RMSE модели для региона

In [12]:
#Вытащим и запишим в переменные выводимые функцией значения для региона №2
model_geo_2 = result_model_geo_2[0]                          #Обученная модель для региона 
predictions_valid_geo_2 = result_model_geo_2[1]              #Предсказания модели на валидационной выборке для региона 
target_valid_geo_2 = result_model_geo_2[2]                   #Правильные ответы валидационной выборке для региона 
mean_predictions_valid_volume_geo_2 = result_model_geo_2[3]  #Средний запас предсказанного сырья для региона
rmse_geo_2 = result_model_geo_2[4]                           #RMSE модели для региона

In [13]:
#Вытащим и запишим в переменные выводимые функцией значения для региона №3
model_geo_3 = result_model_geo_3[0]                          #Обученная модель для региона 
predictions_valid_geo_3 = result_model_geo_3[1]              #Предсказания модели на валидационной выборке для региона
target_valid_geo_3 = result_model_geo_3[2]                   #Правильные ответы валидационной выборке для региона 
mean_predictions_valid_volume_geo_3 = result_model_geo_3[3]  #Средний запас предсказанного сырья для региона
rmse_geo_3 = result_model_geo_3[4]                           #RMSE модели для региона

#### Напечатаем на экране средний запас предсказанного сырья и RMSE модели по регионам. Проанализируем результаты.

In [14]:
#Для региона №1
print('Средний запас предсказанного сырья региона №1:')
print(mean_predictions_valid_volume_geo_1)
print('RMSE модели региона №1:')
print(rmse_geo_1)

Средний запас предсказанного сырья региона №1:
92.56773277214636
RMSE модели региона №1:
37.70188872577139


#### Выводы:
- Регион №1:
   - Средний запас предсказанного сырья: 92.57 тыс. баррелей.
   - RMSE модели: 37.70 тыс. баррелей.
- Модель для региона №1 имеет средний запас предсказанного сырья около 92.57 тыс. баррелей. Однако, с учетом значения RMSE в 37.70 тыс. баррелей, модель может иметь значительную погрешность в предсказаниях.

In [15]:
#Для региона №2
print('Средний запас предсказанного сырья региона №2:')
print(mean_predictions_valid_volume_geo_2)
print('RMSE модели региона №2:')
print(rmse_geo_2)

Средний запас предсказанного сырья региона №2:
68.81242584773999
RMSE модели региона №2:
0.8914237412897414


#### Выводы:
- Регион №2:
  - Средний запас предсказанного сырья: 68.81 тыс. баррелей.
  - RMSE модели: 0.89 тыс. баррелей.
- Для региона №2 средний запас предсказанного сырья составляет около 68.81 тыс. баррелей. Значение RMSE, равное 0.89 тыс. баррелей, указывает на то, что модель обладает низкой ошибкой предсказаний.


In [16]:
#Для региона №3
print('Средний запас предсказанного сырья региона №3:')
print(mean_predictions_valid_volume_geo_3)
print('RMSE модели региона №3:')
print(rmse_geo_3)

Средний запас предсказанного сырья региона №3:
94.94516837037592
RMSE модели региона №3:
40.03777476932527


#### Выводы:
- Регион №3:
  - Средний запас предсказанного сырья: 94.95 тыс. баррелей.
  - RMSE модели: 40.04 тыс. баррелей.
- Для региона №3 средний запас предсказанного сырья примерно равен 94.95 тыс. баррелей, однако, значение RMSE составляет 40.04 тыс. баррелей, что говорит о том, что модель может допускать значительные ошибки в предсказаниях.

## Подготовка к расчёту прибыли

#### Запишем все ключевые значения для расчётов в отдельные переменные.

- При разведке региона исследуют 500 точек, из которых с помощью машинного обучения выбирают 200 лучших для разработки.
- Бюджет на разработку скважин в регионе — 10 млрд рублей.
- При нынешних ценах один баррель сырья приносит 450 рублей дохода. Доход с каждой единицы продукта составляет 450 тыс. рублей, поскольку объём указан в тысячах баррелей.
- После оценки рисков нужно оставить лишь те регионы, в которых вероятность убытков меньше 2.5%. Среди них выбирают регион с наибольшей средней прибылью.

In [17]:
#Количество точек при разведке региона
EXPLORATION_POINTS = 500

#Количество лучших точек для разработки
DEVELOPMENT_POINTS= 200

#Бюджет на разработку скважин в регионе, в рублях
BUDGET= 10_000_000_000 # 10 млрд рублей

#Цена одного барреля сырья, в рублях
BARREL_PRICE = 450

#Доход с каждой единицы продукта, в рублях
INCOME_PER_UNIT = BARREL_PRICE * 1000  # 450 тыс. рублей

#Допустимая вероятность убытков
RISK_THRESHOLD = 0.025

#### Расчет достаточного объема сырья для безубыточной разработки
- Для безубыточной разработки необходимо, чтобы доход от добычи сырья был равен или превышал затраты на разработку скважины. Рассчитаем минимальный объем сырья, который необходим для этого.

In [50]:
#Проведем расчет минимального объема сырья для безубыточной разработки
MIN_VOLUME = (BUDGET/200) / INCOME_PER_UNIT #Поделим бюджет на 200-количество разрабатываемых 
                                            #скважин в регионе и затем поделим на доход с каждой единицы продукта,
                                            #в рублях 

#### Сравним объем сырья для безубыточной работы со средними запасами по регионам.

In [29]:
#Для региона №1
print('Средний запас региона №1 тыс.баррелей:')
print(df_geo_1['product'].mean())
print()
print('Минимальный объем сырья для безубыточной разработки тыс.баррелей:')
print(MIN_VOLUME)

Средний запас региона №1 тыс.баррелей:
92.50000000000001

Минимальный объем сырья для безубыточной разработки тыс.баррелей:
111.11111111111111


In [30]:
#Для региона №2
print('Средний запас региона №2 тыс.баррелей:')
print(df_geo_2['product'].mean())
print()
print('Минимальный объем сырья для безубыточной разработки тыс.баррелей:')
print(MIN_VOLUME)

Средний запас региона №2 тыс.баррелей:
68.82500000000002

Минимальный объем сырья для безубыточной разработки тыс.баррелей:
111.11111111111111


In [31]:
#Для региона №3
print('Средний запас региона №3 тыс.баррелей:')
print(df_geo_3['product'].mean())
print()
print('Минимальный объем сырья для безубыточной разработки тыс.баррелей:')
print(MIN_VOLUME)

Средний запас региона №3 тыс.баррелей:
95.00000000000004

Минимальный объем сырья для безубыточной разработки тыс.баррелей:
111.11111111111111


In [54]:
print('В среднем запас месторождений по регионам отличается от минимального объема сырья для безубыточной разработки на')
print(round((df_geo_1['product'].mean() + df_geo_2['product'].mean() + df_geo_3['product'].mean())/3 - MIN_VOLUME, 2), 'тыс.баррелей')

В среднем запас месторождений по регионам отличается от минимального объема сырья для безубыточной разработки на
-25.67 тыс.баррелей


#### Вывод:

- В каждом регионе средний запас месторождений меньше минимальног объема сырья для безубыточной разработки.
- Видимо как раз из-за идет и есть необходимость в выборе 200 наиболее "богатых" серьем месторождений с помощью машинного обучения.
- В среднем запас месторождений по регионам меньше минимального объема сырья для безубыточной разработки на 25 тыс.баррелей.
- Идем дальше выполнять задачу!

#### Выводы по этапу подготовки расчета прибыли:

- Мы рассмотрели основные параметры и условия задачи: количество исследуемых точек, бюджет на разработку, цену барреля сырья и доход с единицы продукта.
- Рассчитали минимальный объем сырья для безубыточной разработки новой скважины. Этот объем будет являться пороговым значением, которое необходимо превысить для прибыльной разработки.
- Далее мы будем использовать эти ключевые значения для расчета прибыли и оценки рисков в каждом регионе.
- Переходим к расчету прибыли и оценке рисков в каждом регионе.

## Расчёт прибыли и рисков 

#### ~~Напишем функцию для расчёта прибыли по выбранным скважинам и предсказаниям модели.~~

In [59]:
def calculate_profit(target, predictions, development_points, barrel_price, budget, income_per_unit):
    #Преобразуем массива NumPy в Series Pandas
    predictions_series = pd.Series(predictions)

    #Получим предсказания для выбранных индексов
    exploration_points_predictions = predictions.sort_values().tail(development_points)
    #Рассчитаем прибыль для 200 наилучших месторождений из выбранных
    profit = (target.iloc[exploration_points_predictions.index].sum() * income_per_unit) - budget
    
    return profit

#### Напишим функцию для подсчета рисков и прибыли.
- Функция будет применять технику Bootstrap с 1000 выборками. 
- Считать прибыль с помощью функции calculate_profit, написанной выше.
- Находить среднюю прибыль.
- Находить 95%-й доверительный интервал.
- Считать риск убытков.

In [60]:
def calculate_with_bootstrap_risk_and_profit(target, predictions, exploration_points, development_points, barrel_price, budget, income_per_unit, bootstrap_samples=1000):
    #Создадим список с профитами
    profits = []
    #Установим seed для генератора случайных чисел для воспроизводимости результатов
    state = np.random.RandomState(12345)
    #Преобразуем массива NumPy в Series Pandas
    predictions = pd.Series(predictions)
    
    #Напишем цикл для Bootstrap с 1000 выборками
    for sample in range(bootstrap_samples):
        #Выберем случайные месторождения для разведки региона
        exploration_points_random_sample = predictions.sample(n=exploration_points, 
                                                              random_state=state,
                                                              replace = True)
        
        #Рассчитаем прибыль для 200 наилучших месторождений из выбранных
        profit = calculate_profit(target, 
                                  exploration_points_random_sample, 
                                  development_points, 
                                  barrel_price, budget, 
                                  income_per_unit)
        
        #Добавим значение в список с профитами
        profits.append(profit)
    
    #Запишем список профитов как Pandas Series для удобства
    profits = pd.Series(profits)
    profits = profits.astype(float)
    
    #Найдем среднюю прибыль
    mean_profit = profits.mean()
    
    #Найдем 95%-й доверительный интервал
    lower_bound = np.percentile(profits, 2.5)
    upper_bound = np.percentile(profits, 97.5)
    
    #Найдем риск убытков
    loss_probability = (profits < 0).mean() * 100

    #Вернем список с профитами, среднюю прибыль, 95%-й доверительный интервал и риск убытков
    return profits, mean_profit, lower_bound, upper_bound, loss_probability

#### Применим нашу функцию поочередно к регионам и сохраним возвращенные ею значения в соответствующии переменные.

In [61]:
#Количество точек при разведке региона
EXPLORATION_POINTS = 500

#Количество лучших точек для разработки
DEVELOPMENT_POINTS= 200

#Бюджет на разработку скважин в регионе, в рублях
BUDGET= 10_000_000_000 # 10 млрд рублей

#Цена одного барреля сырья, в рублях
BARREL_PRICE = 450

#Доход с каждой единицы продукта, в рублях
INCOME_PER_UNIT = BARREL_PRICE * 1000  # 450 тыс. рублей

#Допустимая вероятность убытков
RISK_THRESHOLD = 0.025

In [62]:
#Применим функцию к каждому датасету и запишем выводимые ее значения в переменную для каждого датасета
#результаты работы функции для региона №1
result_risk_and_profit_geo_1 = calculate_with_bootstrap_risk_and_profit(target_valid_geo_1, 
                                                                        predictions_valid_geo_1, 
                                                                        EXPLORATION_POINTS,
                                                                        DEVELOPMENT_POINTS, 
                                                                        BARREL_PRICE, 
                                                                        BUDGET,
                                                                        INCOME_PER_UNIT,
                                                                        bootstrap_samples=1000)
#результаты работы функции для региона №2
result_risk_and_profit_geo_2 = calculate_with_bootstrap_risk_and_profit(target_valid_geo_2, 
                                                                        predictions_valid_geo_2, 
                                                                        EXPLORATION_POINTS,
                                                                        DEVELOPMENT_POINTS, 
                                                                        BARREL_PRICE, 
                                                                        BUDGET,
                                                                        INCOME_PER_UNIT,
                                                                        bootstrap_samples=1000)
#результаты работы функции для региона №3
result_risk_and_profit_geo_3 = calculate_with_bootstrap_risk_and_profit(target_valid_geo_3, 
                                                                        predictions_valid_geo_3, 
                                                                        EXPLORATION_POINTS,
                                                                        DEVELOPMENT_POINTS, 
                                                                        BARREL_PRICE, 
                                                                        BUDGET,
                                                                        INCOME_PER_UNIT,
                                                                        bootstrap_samples=1000)

#### Из переменных с результатами работы функции выше по регионам вытащим каждое значения и запишим в соответствующую переменную

In [63]:
#Вытащим и запишим в переменные выводимые функцией значения для региона №1
profits_geo_1 = result_risk_and_profit_geo_1[0]           #Список с профитами для региона 
mean_profit_geo_1 = result_risk_and_profit_geo_1[1]       #Cредняя прибыль для региона 
lower_bound_geo_1 = result_risk_and_profit_geo_1[2]       #Нижняя часть 95%-й доверительный интервал для региона 
upper_bound_geo_1 = result_risk_and_profit_geo_1[3]       #Верхняя часть 95%-й доверительный интервал для региона 
loss_probability_geo_1 = result_risk_and_profit_geo_1[4]  #Риск убытков для региона 

In [64]:
#Вытащим и запишим в переменные выводимые функцией значения для региона №2
profits_geo_2 = result_risk_and_profit_geo_2[0]           #Список с профитами для региона 
mean_profit_geo_2 = result_risk_and_profit_geo_2[1]       #Cредняя прибыль для региона 
lower_bound_geo_2 = result_risk_and_profit_geo_2[2]       #Нижняя часть 95%-й доверительный интервал для региона 
upper_bound_geo_2 = result_risk_and_profit_geo_2[3]       #Верхняя часть 95%-й доверительный интервал для региона 
loss_probability_geo_2 = result_risk_and_profit_geo_2[4]  #Риск убытков для региона 

In [65]:
#Вытащим и запишим в переменные выводимые функцией значения для региона №3
profits_geo_3 = result_risk_and_profit_geo_3[0]           #Список с профитами для региона 
mean_profit_geo_3 = result_risk_and_profit_geo_3[1]       #Cредняя прибыль для региона 
lower_bound_geo_3 = result_risk_and_profit_geo_3[2]       #Нижняя часть 95%-й доверительный интервал для региона 
upper_bound_geo_3 = result_risk_and_profit_geo_3[3]       #Верхняя часть 95%-й доверительный интервал для региона 
loss_probability_geo_3 = result_risk_and_profit_geo_3[4]  #Риск убытков для региона 

#### Напечатаем на экране для каждого региона:
- среднюю прибыль 
- 95%-й доверительный интервал 
- риск убытков 
#### Проанализируем результаты.

In [66]:
print('Средняя прибыль для региона №1:')
print(mean_profit_geo_1)
print()
print('95%-й доверительный интервал для региона №1:')
print(lower_bound_geo_1, upper_bound_geo_1)
print()
print('Риск убытков для региона №1:')
print(loss_probability_geo_1, '%')

Средняя прибыль для региона №1:
435783092.39059216

95%-й доверительный интервал для региона №1:
-89511204.38335451 914696444.5588712

Риск убытков для региона №1:
5.1 %


In [67]:
print('Средняя прибыль для региона №2:')
print(mean_profit_geo_2)
print()
print('95%-й доверительный интервал для региона №2:')
print(lower_bound_geo_2, upper_bound_geo_2)
print()
print('Риск убытков для региона №2:')
print(loss_probability_geo_2, '%')

Средняя прибыль для региона №2:
456733222.5407958

95%-й доверительный интервал для региона №2:
35251105.561428644 898690529.3957163

Риск убытков для региона №2:
1.5 %


In [68]:
print('Средняя прибыль для региона №3:')
print(mean_profit_geo_3)
print()
print('95%-й доверительный интервал для региона №3:')
print(lower_bound_geo_3, upper_bound_geo_3)
print()
print('Риск убытков для региона №3:', '%')
print(loss_probability_geo_3)

Средняя прибыль для региона №3:
378460655.79005337

95%-й доверительный интервал для региона №3:
-125080586.48930758 869037548.1112261

Риск убытков для региона №3: %
7.3999999999999995


#### Выберем регион с наибольшей средней прибылью среди регионов с вероятностью убытков меньше 2.5%.

In [77]:
print(f"Единственный регион с вероятностью убытков ниже {RISK_THRESHOLD*100}% это регион №2")
print(f"Его наибольшая средняя прибыль {round(mean_profit_geo_2, 0)} рублей")
print(f"А вероятность рисков убытков для региона {loss_probability_geo_2}%")

Единственный регион с вероятностью убытков ниже 2.5% это регион №2
Его наибольшая средняя прибыль 456733223.0 рублей
А вероятность рисков убытков для региона 1.5%


## Общий вывод:

#### Ход исследования:

- Загрузка и подготовка данных:

   - Были импортированы необходимые библиотеки и модули.
   - Файлы были прочитаны и данные сохранены в переменные.
   - Содержимое датасетов было оценено и описано.
   - Данные были обработаны.

- Обучение и проверка модели:
   - Написана функция для построения модели, включающая все необходимые этапы для обучения модели.
   - Эта функция была применена ко всем датасетам, и результаты были записаны в переменные.
   - Из переменных с результатами работы функции по регионам были извлечены значения и записаны в соответствующие переменные.
   - На экране были напечатаны средние запасы предсказанного сырья и RMSE модели по регионам, а также были проанализированы результаты.
  
- Подготовка к расчёту прибыли:
   - Все ключевые значения для расчётов были записаны в отдельные переменные.
   - Был рассчитан достаточный объём сырья для безубыточной разработки.
   - Было произведено сравнение объёма сырья для безубыточной работы со средними запасами по регионам.

- Расчёт прибыли и рисков:
   - Написана функция для подсчёта рисков и прибыли.
   - Эта функция была применена поочерёдно к регионам, и результаты были сохранены в соответствующих переменных.
   - Из переменных с результатами работы функции по регионам были извлечены значения и записаны в соответствующие переменные.
   - На экране были напечатаны для каждого региона: средняя прибыль, 95%-й доверительный интервал и риск убытков. Результаты были проанализированы.
  
- Выбор региона для разработки:
   - Был выбран регион с наибольшей средней прибылью среди регионов, у которого вероятность убытков меньше 2.5%.
  
### Итог:
- #### Наибольшая средняя прибыль среди регионов с риском убытков меньше 2.5% у региона №2. Она имеет значение: 2 456 733 223 рублей. Его риск убытков — 2.5%. Рекомендуется выбрать именно этот регион для разработки месторождений.

---

## Overall Conclusion:

#### Research Process:

- Data loading and preparation:

   - The necessary libraries and modules were imported.
   - The files were read, and the data was stored in variables.
   - The contents of the datasets were examined and described.
   - The data was processed.

- Model training and evaluation:
   - A function was created to build the model, including all necessary training steps.
   - This function was applied to all datasets, and the results were stored in variables.
   - The relevant values from the regional results were extracted and saved into separate variables.
   - The average predicted reserve volume and the model RMSE for each region were printed and analyzed.
  
- Preparation for profit calculation:
   - All key calculation parameters were stored in separate variables.
   - The minimum reserve volume required for break-even development was calculated.
   - This break-even threshold was compared with the average reserves in each region.

- Profit and risk calculation:
   - A function was created to calculate profit and risks.
   - This function was applied to each region in turn, and the results were stored in corresponding variables.
   - The required values from the regional results were extracted and saved separately.
   - For each region, the average profit, 95% confidence interval, and risk of losses were printed and analyzed.
  
- Region selection:
   - The region with the highest average profit among those with a loss probability below 2.5% was selected.
  
### Final Result:
- #### The highest average profit among the regions with a loss risk below 2.5% was observed in **Region 2**. Its value is **2,456,733,223 rubles**. Its loss risk is **2.5%**. Therefore, this region is recommended for field development.